README.md
# HTTP Monitoring Tools

HTTP Monitoring Tools provides tools for monitoring availability and
certificate validity of HTTP endpoints:

* `getHealthInfo` returns information about HTTP endpoint's availability, such
as HTTP status code, request duration in milliseconds, is endpoint considered healthy.
* `getSslCertificateInfo` returns basic information about HTTP endpoint's SSL
certificate, such as common name, issued date, expire date, number of valid
days left.

## Installation

Add `http-monitoring-tools` as a dependency:

```
$ npm install http-monitoring-tools --save # npm i -s http-monitoring-tools
```

## Usage

### getHealthInfo

CommonJS:
```
const tools = require('http-monitoring-tools')
```

ES module:
```
import tools from 'http-monitoring-tools'
```

Request:
```
tools.getHealthInfo('www.google.com').then((res) => {
  console.log(res)
}).catch((e) =>  {
  console.log(e)
})
```

Response:
```
{
  host: 'www.google.com',
  httpStatusCode: 200,
  isHealthy: true,
  now: '2020-04-26T11:43:23.504Z',
  requestDuration: 153,
  useHttps: true
}
```

`isHealthy` is `true` when `httpStatusCode` is between 200 and 399. Otherwise
`isHealthy` is `false`. Some sites, e.g. sites hosted on Squarespace, may
return status code 400 if `user-agent` HTTP header is missing even if the site
is working normally. Therefore, the default value `{ 'user-agent': '...' }` is
used.

#### Options
| Option | Default | |
|--|--|--|
| port | `443` | Target port |
| path | ` ` | Target path |
| method | `HEAD` | HTTP request method |
| headers | `{ 'user-agent': '...' }` | HTTP request headers |
| https | `true` | Send request using `https`. When `https=false`, `http` is used  |
| timeout | `20000` | HTTP request timeout in milliseconds (`20s` by default) |

Override default options:
```
# Override port and method
tools.getHealthInfo('www.google.com', { port: 8443, method: 'POST' })

# Use http instead of https
tools.getHealthInfo('www.google.com', { port: 80, https: false })

# Override headers
tools.getHealthInfo('www.google.com', { headers: { 'custom-header':'value', 'another-header':'foo'} })
```
### getSslCertificateInfo

CommonJS:
```
const tools = require('http-monitoring-tools')
```

ES module:
```
import tools from 'http-monitoring-tools'
```

Request:
```
tools.getSslCertificateInfo('www.google.com').then((res) => {
  console.log(res)
}).catch((e) =>  {
  console.log(e)
})
```

Response:
```
{
  host: 'www.google.com',
  commonName: 'www.google.com',
  issuerCommonName: 'GTS CA 1O1',
  validFrom: '2020-04-07T09:49:21.000Z',
  validTo: '2020-06-30T09:49:21.000Z',
  now: '2020-04-26T08:09:40.576Z',
  validDaysLeftCount: 65,
  valid: true,
  fingerprint: '51:E2:E8:B8:38:39:0F:FD:8F:5D:3F:93:F4:AE:BC:42:8B:3E:77:13'
}
```

In case connecting to the target succeeds, but reading TLS connection metadata
fails, a less detailed response is returned:

```
{
  host: 'example.com',
  now: '2020-05-03T12:30:30.975Z',
  error: 'Unable to read SSL certificate'
}
```

**N.B.** Self-signed certificates are supported.

#### Options
| Option | Default | |
|--|--|--|
| port | `443` | Target port |
| path | ` ` | Target path |
| method | `HEAD` | HTTP request method |
| headers | `{ 'user-agent': '...' }` | HTTP request headers |
| timeout | `20000` | HTTP request timeout in milliseconds (`20s` by default) |

Override default options:
```
# Override port and method
tools.getSslCertificateInfo('www.google.com', { port: 8443, method: 'POST' })

# Override headers
tools.getSslCertificateInfo('www.google.com', { headers: { 'custom-header':'value'} })
```

## Build

```
$ git clone https://github.com/petkivim/http-monitoring-tools.git
$ npm install
$ npm run cjs
```


HISTORY.md
1.2.0 / 2022-04-16
==================

  * Disable connection pooling
  * Return an error object if reading TLS connection metadata fails
  * Change connection timeout default from 30s to 20s
  * Make connection timeout configurable
  * Update dependencies

1.1.0 / 2020-04-27
==================

  * Support for CommonJS and ES module

1.0.0 / 2020-04-26
==================

  * Initial release


http-monitoring-tools.js
/**
 * The MIT License
 *
 * Copyright (c) 2020 Petteri Kivimäki
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
import https from 'https'

// One day in milliseconds
const millisPerDay = 1000 * 60 * 60 * 24

/*
    Calculate difference between the dates, divide by milliseconds per day,
    and round downwards to nearest integer.
  */
function dateDiff(firstDate, secondDate) {
  return Math.floor((new Date(secondDate) - new Date(firstDate)) / (millisPerDay))
}

/*
   Calculate the number of days between the given date and current date. If
   the given date is in the past, the returned value is negative.
 */
function getValidDaysCount(validTo) {
  return dateDiff(new Date(), validTo)
}

/*
  Return the given date as ISO 8601 formatted string:
  YYYY-MM-DDTHH:mm:ss.sssZ
 */
function formatDate(date) {
  return new Date(date).toISOString()
}

/*
  Generate response object returned by the getSslCertificateInfo function.
 */
function generateSslCertificateInfoResponse(res, options) {
  const validDaysCount = getValidDaysCount(res.certificate.valid_to)
  return {
    host: options.hostname,
    commonName: res.certificate.subject.CN,
    issuerCommonName: res.certificate.issuer.CN,
    validFrom: formatDate(res.certificate.valid_from),
    validTo: formatDate(res.certificate.valid_to),
    now: formatDate(new Date()),
    validDaysLeftCount: validDaysCount,
    valid: validDaysCount >= 0,
    fingerprint: res.certificate.fingerprint,
  }
}

/*
  Generate options for getSslCertificateInfo function by merging user
  supplied options and default options.
*/
function getSslCertificateInfoOptions(hostname, options = {}) {
  const headers = { 'user-agent': '...' }
  return {
    hostname,
    port: options.port === undefined ? 443 : options.port,
    path: options.path === undefined ? '' : options.path,
    method: options.method === undefined ? 'HEAD' : options.method,
    // Self-signed certificates are supported
    rejectUnauthorized: false,
    // Connection timeout is 30s
    timeout: 30000,
    headers: options.headers === undefined ? headers : options.headers,
  }
}

export default {
  getSslCertificateInfo(hostname, userOptions) {
    return new Promise((resolve, reject) => {
      const options = getSslCertificateInfoOptions(hostname, userOptions)
      const req = https.request(options, (res) => {
        res.certificate = res.connection.getPeerCertificate()
        resolve(generateSslCertificateInfoResponse(res, options))
      }).on('timeout', () => {
        req.abort()
      }).on('error', (e) => {
        reject(e)
      })
      req.end()
    })
  },
}


esm/index.js
/**
 * The MIT License
 *
 * Copyright (c) 2020 Petteri Kivimäki
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
import http from 'http'
import https from 'https'

// One day in milliseconds
const millisPerDay = 1000 * 60 * 60 * 24
// Default request headers
const defaultHeaders = { 'user-agent': '...' }

/*
    Calculate difference between the dates, divide by milliseconds per day,
    and round downwards to nearest integer.
  */
function dateDiff(firstDate, secondDate) {
  return Math.floor((new Date(secondDate) - new Date(firstDate)) / (millisPerDay))
}

/*
   Calculate the number of days between the given date and current date. If
   the given date is in the past, the returned value is negative.
 */
function getValidDaysCount(validTo) {
  return dateDiff(new Date(), validTo)
}

/*
  Return the given date as ISO 8601 formatted string:
  YYYY-MM-DDTHH:mm:ss.sssZ
 */
function formatDate(date) {
  return new Date(date).toISOString()
}

/*
  Generate response object returned by the getSslCertificateInfo function.
 */
function generateSslCertificateInfoResponse(res, options) {
  const validDaysCount = getValidDaysCount(res.certificate.valid_to)
  try {
    return {
      host: options.hostname,
      commonName: res.certificate.subject.CN,
      issuerCommonName: res.certificate.issuer.CN,
      validFrom: formatDate(res.certificate.valid_from),
      validTo: formatDate(res.certificate.valid_to),
      now: formatDate(new Date()),
      validDaysLeftCount: validDaysCount,
      valid: validDaysCount >= 0,
      fingerprint: res.certificate.fingerprint,
    }
  } catch (e) {
    return {
      host: options.hostname,
      now: formatDate(new Date()),
      error: 'Unable to read SSL certificate',
    }
  }
}

/*
  Generate response object returned by the getHealthInfo function.
 */
function generateHealthInfoResponse(res, options, startAt) {
  const isHealthy = res.statusCode >= 200 && res.statusCode < 400
  return {
    host: options.hostname,
    httpStatusCode: res.statusCode,
    isHealthy,
    now: formatDate(new Date()),
    requestDuration: new Date() - startAt,
    useHttps: options.https,
  }
}

/*
  Generate options for getHealthInfo and getSslCertificateInfo functions
  by merging user supplied options and default options.
*/
function getRequestOptions(hostname, options = {}) {
  return {
    hostname,
    port: options.port === undefined ? 443 : options.port,
    path: options.path === undefined ? '' : options.path,
    method: options.method === undefined ? 'HEAD' : options.method,
    // Self-signed certificates are supported
    rejectUnauthorized: false,
    // Connection timeout is 20s by default
    timeout: options.timeout === undefined ? 20000 : options.timeout,
    // Opt out of connection pooling by not using an Agent
    agent: false,
    headers: options.headers === undefined ? defaultHeaders : options.headers,
    https: options.https === undefined ? true : options.https,
  }
}

export function getSslCertificateInfo(hostname, userOptions) {
  return new Promise((resolve, reject) => {
    const options = getRequestOptions(hostname, userOptions)
    const req = https.request(options, (res) => {
      res.certificate = res.connection.getPeerCertificate()
      resolve(generateSslCertificateInfoResponse(res, options))
    }).on('timeout', () => {
      req.abort()
    }).on('error', (e) => {
      reject(e)
    })
    req.end()
  })
}

export function getHealthInfo(hostname, userOptions) {
  return new Promise((resolve, reject) => {
    const options = getRequestOptions(hostname, userOptions)
    const startAt = new Date()
    const req = (options.https ? https : http).request(options, (res) => {
      resolve(generateHealthInfoResponse(res, options, startAt))
    }).on('timeout', () => {
      req.abort()
    }).on('error', (e) => {
      reject(e)
    })
    req.end()
  })
}

export default { getHealthInfo, getSslCertificateInfo }
